cmd/controller/main.go

```diff
@@ -16,6 +23,52 @@ import (
 // Version of kapp-controller is set via ldflags at build-time from the most recent git tag; see hack/build.sh
 var Version = "develop"
 
+var (
+	client *clientset.Clientset
+)
+
+func getNewLock(lockname, podname, namespace string) *resourcelock.LeaseLock {
+	return &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      lockname,
+			Namespace: namespace,
+		},
+		Client: client.CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: podname,
+		},
+	}
+}
+
+func runLeaderElection(lock *resourcelock.LeaseLock, ctx context.Context, id string, ctrlOpts Options, mainLog logr.Logger, log logr.Logger) {
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock:            lock,
+		ReleaseOnCancel: true,
+		LeaseDuration:   15 * time.Second,
+		RenewDeadline:   10 * time.Second,
+		RetryPeriod:     2 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(c context.Context) {
+				err := Run(ctrlOpts, log.WithName("controller"))
+				if err != nil {
+					mainLog.Error(err, "Exited run with error")
+					os.Exit(1)
+				}
+			},
+			OnStoppedLeading: func() {
+				klog.Info("no longer the leader, staying inactive.")
+			},
+			OnNewLeader: func(current_id string) {
+				if current_id == id {
+					klog.Info("still the leader.")
+					return
+				}
+				klog.Info("new leader is %s", current_id)
+			},
```

Comment:
shouldnt these methods be doing something to avoid having "old" kapp-controller continuing to execute?